### PR TITLE
Implement token revoke KYC handle

### LINKED ISSUE #5912: TokenRevokeKycFromAccountHandler

### PR DESCRIPTION
**Pull request by mhess-swl**
Closes #5912

### CODE REVIEW

**Review by hendrikebbers — DISMISSED** [review]

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/dispatcher/TransactionDispatcher.java [hunk 1]
```diff
@@ -293,6 +295,19 @@ private void dispatchTokenGrantKycToAccount(
...
     }
 
+    /**
+     * Dispatches the token revoke KYC transaction to the appropriate handler.
+     *
+     * @param tokenRevokeKyc the token revoke KYC transaction
+     * @param tokenRelStore the token relation store
+     */
+    private void dispatchTokenRevokeKycFromAccount(
+            TransactionBody tokenRevokeKyc, WritableTokenRelationStore tokenRelStore) {
```
Comment (hendrikebbers) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/dispatcher/TransactionDispatcher.java:305]: final and null annotations missing

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/dispatcher/TransactionDispatcher.java [hunk 2]
```diff
@@ -329,12 +331,25 @@ protected void finishConsensusSubmitMessage(
...
+     * Dispatches the token revoke KYC transaction to the appropriate handler.
+     *
+     * @param tokenRevokeKyc the token revoke KYC transaction
+     * @param tokenRelStore the token relation store
+     */
+    private void dispatchTokenRevokeKycFromAccount(
+            @NonNull TransactionBody tokenRevokeKyc, @NonNull WritableTokenRelationStore tokenRelStore) {
+        final var handler = handlers.tokenRevokeKycFromAccountHandler();
+        handler.handle(tokenRevokeKyc, tokenRelStore);
+        tokenRelStore.commit();
```
Comment (tinker-michaelj) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/dispatcher/TransactionDispatcher.java:350]: These `*Store.commit()` calls will be done by the workflow somewhere else eventually, so atm are just a `mono-service`-adapter pattern.

So we might move this call into a `MonoTransactionDispatcher#finishTokenRevokeKycFromAccount()` method a la [here](https://github.com/hashgraph/hedera-services/blob/f80240f84ac872a9590e0eaf286c310509cb1507/hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/dispatcher/MonoTransactionDispatcher.java#L61).

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenRevokeKycFromAccountHandler.java [hunk 3]
```diff
@@ -60,7 +63,7 @@ public void preHandle(@NonNull final PreHandleContext context, @NonNull final Re
         requireNonNull(context);
         final var op = context.body().tokenRevokeKycOrThrow();
         final var tokenMeta = tokenStore.getTokenMeta(op.tokenOrElse(TokenID.DEFAULT));
-        if (tokenMeta == null) throw new PreCheckException(INVALID_TOKEN_ID);
+        if (tokenMeta == null) throw new PreCheckException(ResponseCodeEnum.INVALID_TOKEN_ID);
         if (tokenMeta.hasKycKey()) {
```
Comment (tinker-michaelj) [inline hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenRevokeKycFromAccountHandler.java:67]: ⚠️ Don't we need to fail here with `TOKEN_HAS_NO_KYC_KEY` if the KYC key is missing?

**Review by tinker-michaelj — CHANGES_REQUESTED** [review]
Seems to be missing the `TOKEN_HAS_NO_KYC_KEY` check, but otherwise very sensible! 🥂

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenRevokeKycFromAccountHandler.java (hunk 3 above)
Comment (mhess-swl) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenRevokeKycFromAccountHandler.java:67]: I only added the class name, but I can change this too 👍

**Review by tinker-michaelj — APPROVED** [review]
LGTM!

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenRevokeKycFromAccountHandler.java [hunk 4]
```diff
@@ -40,27 +46,57 @@ public TokenRevokeKycFromAccountHandler() {
...
      * @throws NullPointerException if one of the arguments is {@code null}
      */
-    public void handle() {
-        throw new UnsupportedOperationException("Not implemented");
+    public void handle(@NonNull TransactionBody txn, @NonNull WritableTokenRelationStore tokenRelStore) {
+        requireNonNull(txn);
+        requireNonNull(tokenRelStore);
+
+        final var op = txn.tokenRevokeKycOrThrow();
+        verifyOp(op);
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenRevokeKycFromAccountHandler.java:82]: I am adding this method name as `pureChecks` that validates transaction body and doesn't involve dynamic properties and state . We can decide on one name and should probably use it.

Comment (mhess-swl) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenRevokeKycFromAccountHandler.java:82]: `pureChecks` works for me 👍

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenRevokeKycFromAccountHandler.java [hunk 5]
```diff
@@ -40,27 +46,57 @@ public TokenRevokeKycFromAccountHandler() {
...
+        final var tokenId = op.tokenOrThrow().tokenNum();
+        final var accountId = op.accountOrElse(AccountID.DEFAULT).accountNumOrThrow();
+        final var tokenRel = tokenRelStore.getForModify(tokenId, accountId);
+
+        final var tokenRelBuilder = tokenRel.orElseThrow().copyBuilder();
+        tokenRelBuilder.kycGranted(false);
+        tokenRelStore.put(tokenRelBuilder.build());
+    }
+
+    private void pureChecks(TokenRevokeKycTransactionBody op) {
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenRevokeKycFromAccountHandler.java:93]: These should be in `preHandle` and any checks that depend on `dynamicProperties` or state should be added to handle

Comment (mhess-swl) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenRevokeKycFromAccountHandler.java:93]: I'll move `pureChecks` to `preHandle`. Is anything missing from `handle`? I didn't see anything else in the current transition logic

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenRevokeKycFromAccountHandler.java [hunk 6]
```diff
@@ -40,27 +46,57 @@ public TokenRevokeKycFromAccountHandler() {
...
+     * @throws NullPointerException if one of the arguments is {@code null}
+     */
     @Override
     public void preHandle(@NonNull final PreHandleContext context) throws PreCheckException {
         requireNonNull(context);
         final var op = context.body().tokenRevokeKycOrThrow();
         final var tokenStore = context.createStore(ReadableTokenStore.class);
         final var tokenMeta = tokenStore.getTokenMeta(op.tokenOrElse(TokenID.DEFAULT));
-        if (tokenMeta == null) throw new PreCheckException(INVALID_TOKEN_ID);
+        if (tokenMeta == null) throw new PreCheckException(ResponseCodeEnum.INVALID_TOKEN_ID);
```
Comment (rbair23) [inline hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenRevokeKycFromAccountHandler.java:62]: FYI -- in a lot of other places I've seen us switch from ResponseCodeEnum.FOO to just using FOO.

**Review by rbair23 — APPROVED** [review]

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenRevokeKycFromAccountHandler.java (hunk 5 above)
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenRevokeKycFromAccountHandler.java:93]: No handle seems fine for me

**Review by Neeharika-Sompalli — APPROVED** [review]
LGTM ! `pureChecks` can be moved to preHandle

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenRevokeKycFromAccountHandler.java (hunk 6 above)
Comment (mhess-swl) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenRevokeKycFromAccountHandler.java:62]: Good to know. Guess I've been hitting the other option in the intellij dropdown 😂

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenRevokeKycFromAccountHandler.java (hunk 5 above)
Comment (mhess-swl) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenRevokeKycFromAccountHandler.java:93]: This has the needed approvals and it's been open for awhile so I'll go ahead and merge, but I'll get these moved to `preHandle` in a PR today

Comment (mhess-swl) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenRevokeKycFromAccountHandler.java:93]: @Neeharika-Sompalli the move is in PR #6318

**Pull request merged**